client/v3/lease.go

```diff
@@ -295,17 +295,46 @@ func (l *lessor) KeepAlive(ctx context.Context, id LeaseID) (<-chan *LeaseKeepAl
 	l.mu.Unlock()
 
 	go l.keepAliveCtxCloser(ctx, id, ka.donec)
+	errorC := make(chan error, 1)
 	l.firstKeepAliveOnce.Do(func() {
-		go l.recvKeepAliveLoop()
+		go l.recvKeepAliveLoop(errorC)
 		go l.deadlineLoop()
 	})
 
+	select {
+	case err := <-errorC:
+		if err != nil {
+			return ch, err
+		}
+	case _, ok := <-ch:
+		if !ok { // The channel has already been closed by l.deadlineLoop()
+			return ch, rpctypes.ErrTimeoutDueToConnectionLost
+		}
+	}
+
 	return ch, nil
 }
 
 func (l *lessor) KeepAliveOnce(ctx context.Context, id LeaseID) (*LeaseKeepAliveResponse, error) {
+	notifyC := make(chan struct{})
+
 	for {
-		resp, err := l.keepAliveOnce(ctx, id)
+		var (
+			resp *LeaseKeepAliveResponse
+			err  error
+		)
+
+		go func() {
+			resp, err = l.keepAliveOnce(ctx, id)
+			notifyC <- struct{}{}
+		}()
+
+		select {
+		case <-notifyC:
+		case <-time.After(l.firstKeepAliveTimeout):
```

Comment:
This code-path would lead to leaked go-routine. 

Would enhancing by context.WithTimeout (https://pkg.go.dev/context#WithTimeout) work here ? 
I would expect grpc layer to obey it properly.... maybe there is some retry logic that does not obey ctx.Done() contract ?

Reply:
Right. Fixed per your comment.  Verified that it works. Thanks.  @ptabor